src/daemon/watcher.c

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "daemon/common.h"
+#include "watcher.h"
+
+watcher_step_t *watcher_steps;
+
+static struct completion shutdown_begin_completion;
+static struct completion shutdown_end_completion;
+static netdata_thread_t watcher_thread;
+
+void watcher_shutdown_begin(void) {
+    completion_mark_complete(&shutdown_begin_completion);
+}
+
+void watcher_shutdown_end(void) {
+    completion_mark_complete(&shutdown_end_completion);
+}
+
+void watcher_step_complete(watcher_step_id_t step_id) {
+    completion_mark_complete(&watcher_steps[step_id].p);
+}
+
+void *watcher_main(void *arg)
+{
+    UNUSED(arg);
+
+    netdata_log_debug(D_SYSTEM, "Watcher thread started");
+
+    completion_wait_for(&shutdown_begin_completion);
+    usec_t shutdown_start_time = now_monotonic_usec();
+
+    // TODO:
+    //   - add a version of completion_wait_for with timeout
+    //   - check the step's duration and abort when the timeout has expired.
```

Comment:
I suppose this `TODO` will come with another PR, please, can you confirm this?

Reply:
Correct.